[libc++] add shared_recursive_mutex api

Jolyon0202 opened the pull request:
libc++'s shared_mutex favor writers over readers, but libstdc++'s shared_mutex favor readers over writers defaultly.

libstdc++¡®s shared_mutex use ptread_rwlock_t(PTHREAD_RWLOCK_PREFER_READER_NP defaultly), and it also support to favor writers over readers.

Official C++ Standard doesn't specify the std::shared_mutex policy. Explanation could be found in original N2406 proposal (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2406.html#shared_mutex_imp) in shared_mutex.

And we add shared_recursive_mutex api that favor readers over writers, and support Write-read nesting and write nesting.

frederick-vs-ja commented on libcxx/include/shared_recursive_mutex:
```diff
@@ -0,0 +1,61 @@
+    void unlock_shared();
+
+private:
+    std::mutex mutex;
+    std::condition_variable cond;
+    pthread_t owner = 0;     // write owner.
+    uint32_t recursions = 0; // read lock recursions
+    uint32_t readers = 0;
```
Member names should be __uglified to avoid conflict with user-defined macros. Perhaps `mutex` is an exception, but it's confusing that there's both `std::mutex` and member `mutex`.

Jolyon0202 replied:
OK, I add namespace __shared_recurisive_mutex and use member __mutex_.

frederick-vs-ja commented on libcxx/include/shared_recursive_mutex:
```diff
@@ -0,0 +1,61 @@
+private:
+    std::mutex mutex;
+    std::condition_variable cond;
+    pthread_t owner = 0;     // write owner.
```
`__thread_id` (available in [`<__thread/id.h>`](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__thread/id.h)) can be used instead, which is available on all supported platforms.

Jolyon0202 replied:
done.

frederick-vs-ja commented on libcxx/include/shared_recursive_mutex:
```diff
@@ -0,0 +1,61 @@
+#include <thread>
+#include <shared_mutex>
+
+namespace std {
```
This line is confusing. Do we want to put this non-standard extension into `std`? Should we use an __uglified class name (`__shared_recursive_mutex`)?

Jolyon0202 replied:
done

frederick-vs-ja commented on libcxx/include/shared_recursive_mutex:
```diff
@@ -0,0 +1,61 @@
+#ifndef _SHARED_RECURSIVE_MUTEX_
+#define _SHARED_RECURSIVE_MUTEX_
+
+#include <cstdint>
+#include <mutex>
+#include <condition_variable>
+#include <thread>
+#include <shared_mutex>
```
We should use some divided internal headers. I don't see why `<shared_mutex>` should be included here.

```suggestion
#include <__condition_variable/condition_variable.h>
#include <__mutex/mutex.h>
#include <__thread/id.h>
#include <cstdint>
```

frederick-vs-ja commented on libcxx/src/shared_recursive_mutex.cpp:
```diff
@@ -0,0 +1,122 @@
+#include <cassert>
```
`assert` is not used. But it seems that we should include `<__mutex/lock_guard.h>`.

```suggestion
#include <__mutex/lock_guard.h>
```

frederick-vs-ja commented on libcxx/src/shared_recursive_mutex.cpp:
```diff
@@ -0,0 +1,122 @@
+#include <shared_recursive_mutex>
+
+constexpr uint32_t SHARED_USER_MAX_CNT = UINT32_MAX;
+constexpr pthread_t THREAD_ID_NOEXIST  = 0;
```
If it's intended to support non-pthread platforms, `__thread_id` should be used. But it's a bit unfortunate that it's default constructor is not constexpr, and I _think_ we can drop this constant and use `__thread_id{}` instead.

frederick-vs-ja commented on libcxx/src/shared_recursive_mutex.cpp:
```diff
@@ -0,0 +1,122 @@
+
+using namespace std;
+
+#ifndef likely
+#  define likely(condition) __builtin_expect(!!(condition), 1)
+#endif
+#ifndef unlikely
+#  define unlikely(condition) __builtin_expect(!!(condition), 0)
+#endif
```
These don't seem necessary as compilers should support `[[likely]]` and `[[unlikely]]` (`[[__likely__]]` and `[[__unlikely__]]` in old modes).

frederick-vs-ja commented on libcxx/src/shared_recursive_mutex.cpp:
```diff
@@ -0,0 +1,122 @@
+#endif
+
+void shared_recursive_mutex::lock() {
+  pthread_t self = pthread_self();
```
If it's intended to support non-pthread platforms ...
```suggestion
  __thread_id self = this_thead::get_id();
```
(ditto for other `self` definitions)

Jolyon0202 replied:
Thx. I have updated a new version of this patch.

Commit:
[libc++] add shared_recursive_mutex api (#80000)

libc++'s shared_mutex favor writers over readers, but libstdc++'s shared_mut
ex favor readers over writers defaultly.

libstdc++'s shared_mutex use ptread_rwlock_t(PTHREAD_RWLOCK_PREFER_READER
_NP defaultly), and it also support to favor writers over readers.

Official C++ Standard doesn't specify the std::shared_mutex policy. Explanat
ion could be found in original N2406 proposal (http://www.open-std.org/jtc1/sc22
/wg21/docs/papers/2007/n2406.html#shared_mutex_imp) in shared_mutex.

And we add shared_recursive_mutex api that favor readers over writers, and s
upport Write-read nesting and write nesting.

philnik777 commented:
What exactly are you trying to do? We don't want to add non-standard extensions if that is what you try to do here.

philnik777 commented:
Closing, since there hasn't been a response for months.
